Bluetooth: controller: Fix completition of rejected LLCP request

MariuszSkamra opened the pull request:
This fixes LLCP procedure completition in case when remote rejected
LLCP Slave Connection Parameter Request with LMP Error Transaction
Collision error code.

The link was disconnected after 40 seconds after receiving the
LLCP Reject Extended Indication of Connection Parameter Update
with the reason LMP Error Transaction Collision.
This patch resets the timer upon reject is received.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -1982,11 +1982,17 @@ isr_rx_conn_pkt_ctrl_rej_conn_upd(struct radio_pdu_node_rx *node_rx,
 		}
 	}
 
+	/* Procedure complete */
```
As it was a LL same procedure collision, the local device would keep the CPR context and timeout. The peer shall complete the connection parameter request procedure hence it is not required to reset the state and timeout.

Could you give details of the issue leading to this PR?

cvinayak replied:
Analyzed the sniffer logs. This PR is to workaround peer implementation that dont consider at Connection Update Procedure (CUP) instant the procedure being complete.

Hence move the code from line 1971-1977 (reset ack = req and procedure_expire = 0) to line 1985 here.

MariuszSkamra replied:
done

cvinayak commented:
From the sniffer log analysis, the previous connection parameter request procedure is complete at the instant, it should not be rejecting the next connection parameter request PDU from the slave.

The peer implementation is not following the BT Spec correctly on two counts:
1. At the instant the connection parameter request procedure is complete.
2. The master shall proceed with its initiated connection parameter request procedure if it returned LL Procedure Collision (0x23) which clearly is wrong as it did not initiated the previous procedure but was initiated by slave.

![image](https://user-images.githubusercontent.com/6350656/56011436-e3c43c00-5ce7-11e9-8211-71545b793535.png)

![image](https://user-images.githubusercontent.com/6350656/56011519-4f0e0e00-5ce8-11e9-9191-a03c3cfb110c.png)

As a workaround to the issue, please update the commit title and commit message accordingly.

"Bluetooth: controller: Workaround CPR procedure collision at CPU instant"

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -1982,11 +1982,17 @@ isr_rx_conn_pkt_ctrl_rej_conn_upd(struct radio_pdu_node_rx *node_rx,
 		return;
 	}
 
+	/* Stop procedure timeout */
```
Remove this.

MariuszSkamra commented:
@cvinayak done. Do you have a possibility to run this against EBQ?
This PR is for v1.13-branch. Do you want to cherry-pick it to master?

cvinayak commented:
@MariuszSkamra yes. Running it, will comment here on the outcome once the testing completes. Yes, please send a PR to master, will approve it.

cvinayak commented:
Run #17041 - /LL/CON/SLA/BV-26-C	Initiating Connection Parameter Request - same procedure collision
Failed for the change in this PR. Will investigate and get back.

cvinayak commented:
Please try the attached patch, if ok, update this PR.

[0001-Bluetooth-controller-Workaround-CPR-procedure-collis.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/3114658/0001-Bluetooth-controller-Workaround-CPR-procedure-collis.patch.txt)

MariuszSkamra committed:
Bluetooth: controller: Workaround CPR procedure collision at CPU instant

This is a workaround for IOP issue, where peer rejects LLCP Slave
Connection Parameter Request with LMP Error Transaction Collision
error code even if previous request is complete at the instant.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

MariuszSkamra commented:
@cvinayak Thank You! The patch you provided resolves my issue ^^

joerchan commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -1957,29 +1957,27 @@ isr_rx_conn_pkt_ctrl_rej_conn_upd(struct radio_pdu_node_rx *node_rx,
+		}
+	}
+
+	/* Same or Different Procedure Collision */
```
Comment is not related to the if-statement anymore. At least it does not make sense anymore. Update or delete?

MariuszSkamra replied:
deleted
